### ISSUE TITLE
Start background prefetching without calling next on dataset iterator

### ISSUE THREAD

**Issue by Tilps** [tensorflow/tensorflow#35080]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Currently if you have a dataset with a large shuffle buffer, it doesn't start trying to populate that buffer until the first request to retrieve data from the data set. This is a lost opportunity for parallelism during training script startup - the shuffle buffer could be filling in the background while the model is compiling.  
Since calling next is blocking, its not possible to trigger this overlap to happen. 

**Will this change the current api? How?** Maybe - could add a start_prefetching() method.  The other idea I had was to make it implicit in calling __iter__ - but that is possibly more useful for my specific use case than more generally.

**Who will benefit with this feature?**
Anyone who cares about the startup time of their training script and uses a large shuffle buffer.

**Any Other info.**

**Comment by jsimsa** [comment]
I agree that it would be generally beneficial to be able to overlap the "warming up" of the input pipeline with other computation (such as model compilation).

However, it is non-trivial to provide this functionality given the current tf.data implementation. Each tf.data transformation has its own C++ iterator implementation with a "GetNext" method which upon its first invocation starts the background activity (such as prefetching). To add support for "start prefetching" we would need to add a new method to all C++ iterator implementations which starts the background activity (if the iterator is asynchronous) or recursively invokes this method on the input iterators (if the iterator is synchronous).

As an additional concern, if the input pipeline used custom tf.data transformations (e.g. from [SIG I/O](https://github.com/tensorflow/io)) they would also need to implement this method in order for the "start prefetch" functionality to work correctly across all input pipelines.

All in all, I think that the current implementation prevents the functionality to be implemented in an effective and robust manner and for that reason, the tf.data team does not plan to support such functionality.

PS: If you are writing your own custom training loop (and calling `iter` and `next` explicitly), then you should be able to use [tf.group](https://www.tensorflow.org/api_docs/python/tf/group?version=stable) or [tf.tuple](https://www.tensorflow.org/api_docs/python/tf/tuple?version=stable) + `tf.py_function` to wrap the model compilation to parallelize the model compilation and fetching of the first element.

**Comment by Tilps** [comment]
Thanks for the detailed analysis.

Just to check - did you consider a variant where the prefetch transform takes an optional parameter 'eager prefetch'?  If the prefetch transform decouples things, (as was my understanding without having dug into the code) - when provided the eager prefetch flag it would invoke the standard GetNext in the background, in advance of the first attempt to invoke iter/GetNext on the prefetch itself.  This doesn't seem to need an extensive API redesign - although it does assume a lot about  how prefetch works.

**Comment by jsimsa** [comment]
That's an interesting idea. `prefetch` is not the only asychronous transformation though (parallel map and parallel interleave are examples of others).

Having said that, I can imagine having a tf.data.Option that specify whether input pipeline should start background threads eagerly (similar to how you can specify whether input pipeline is allowed to execute in a non-deterministic fashion). This option would then trigger a graph rewrite at input pipeline instantiation time, which would set the execution mode for all asynchronous ops in the input pipeline.

In other words, this could work. I will re-open this issue and mark it as contributions welcome as the tf.data team will not have cycles to prioritize this in Q1.
